File: elixir/apps/arc_data/lib/arc/data/frame.ex
defmodule Arc.Data.Frame do
  @moduledoc """
  Binary request/response frame for ARC handler RPC.

  Wire format:

      [1 byte]   version
      [1 byte]   type
                 1=request
                 2=response
                 3=error
                 4=stream_open
                 5=stream_data
                 6=stream_resize
                 7=stream_close
                 8=stream_exit
                 9=stream_error
                 10=event (server-initiated, meta.topic names it)
      [2 bytes]  flags
      [16 bytes] request_id
      [4 bytes]  meta_length (big-endian)
      [N bytes]  meta JSON
      [4 bytes]  body_length (big-endian)
      [N bytes]  body bytes
  """

  @version 1
  @request_type 1
  @response_type 2
  @error_type 3
  @stream_open_type 4
  @stream_data_type 5
  @stream_resize_type 6
  @stream_close_type 7
  @stream_exit_type 8
  @stream_error_type 9
  @event_type 10
  @request_id_bytes 16

  @type frame_type ::
          :request
          | :response
          | :error
          | :stream_open
          | :stream_data
          | :stream_resize
          | :stream_close
          | :stream_exit
          | :stream_error
          | :event

  @type decoded_frame :: %{
          version: pos_integer(),
          type: frame_type(),
          flags: non_neg_integer(),
          request_id: binary(),
          meta: map(),
          body: binary()
        }

  @doc """
  Create a new random request id.
  """
  @spec new_request_id() :: binary()
  def new_request_id do
    :crypto.strong_rand_bytes(@request_id_bytes)
  end

  @doc """
  Encode a request frame.
  """
  @spec encode_request(binary(), map(), binary()) :: binary()
  def encode_request(request_id, meta, body) when is_map(meta) and is_binary(body) do
    encode_frame(:request, request_id, meta, body)
  end

  @doc """
  Encode a response frame.
  """
  @spec encode_response(binary(), map(), binary()) :: binary()
  def encode_response(request_id, meta, body) when is_map(meta) and is_binary(body) do
    encode_frame(:response, request_id, meta, body)
  end

  @doc """
  Encode an error frame.
  """
  @spec encode_error(binary(), String.t(), String.t(), map(), binary()) :: binary()
  def encode_error(request_id, code, message, meta \\ %{}, body \\ "")
      when is_binary(code) and is_binary(message) and is_map(meta) and is_binary(body) do
    error_meta = Map.merge(%{"code" => code, "message" => message}, meta)
    encode_frame(:error, request_id, error_meta, body)
  end

  @spec encode_stream_open(binary(), map(), binary()) :: binary()
  def encode_stream_open(request_id, meta, body) when is_map(meta) and is_binary(body) do
    encode_frame(:stream_open, request_id, meta, body)
  end

  @spec encode_stream_data(binary(), map(), binary()) :: binary()
  def encode_stream_data(request_id, meta, body) when is_map(meta) and is_binary(body) do
    encode_frame(:stream_data, request_id, meta, body)
  end

  @spec encode_stream_resize(binary(), map(), binary()) :: binary()
  def encode_stream_resize(request_id, meta, body \\ "") when is_map(meta) and is_binary(body) do
    encode_frame(:stream_resize, request_id, meta, body)
  end

  @spec encode_stream_close(binary(), map(), binary()) :: binary()
  def encode_stream_close(request_id, meta, body \\ "") when is_map(meta) and is_binary(body) do
    encode_frame(:stream_close, request_id, meta, body)
  end

  @spec encode_stream_exit(binary(), map(), binary()) :: binary()
  def encode_stream_exit(request_id, meta, body \\ "") when is_map(meta) and is_binary(body) do
    encode_frame(:stream_exit, request_id, meta, body)
  end

  @spec encode_stream_error(binary(), String.t(), String.t(), map(), binary()) :: binary()
  def encode_stream_error(request_id, code, message, meta \\ %{}, body \\ "")
      when is_binary(code) and is_binary(message) and is_map(meta) and is_binary(body) do
    error_meta = Map.merge(%{"code" => code, "message" => message}, meta)
    encode_frame(:stream_error, request_id, error_meta, body)
  end

  @doc """
  Encode an event frame: a message a provider sends without a request.
  `topic` names the event. The request id is fresh; nothing correlates it.
  """
  @spec encode_event(String.t(), binary(), map()) :: binary()
  def encode_event(topic, body, meta \\ %{}) when is_binary(topic) and is_binary(body) do
    encode_frame(:event, new_request_id(), Map.put(meta, "topic", topic), body)
  end

  @spec encode_frame(frame_type(), binary(), map(), binary()) :: binary()
  def encode_frame(type, request_id, meta, body)
      when is_atom(type) and is_map(meta) and is_binary(body) do
    encode(type_code(type), request_id, meta, body)
  end

  defp encode(type, request_id, meta, body)
       when type in [
              @request_type,
              @response_type,
              @error_type,
              @stream_open_type,
              @stream_data_type,
              @stream_resize_type,
              @stream_close_type,
              @stream_exit_type,
              @stream_error_type,
              @event_type
            ] do
    req_id = ensure_request_id(request_id)
    meta_bytes = IO.iodata_to_binary(:json.encode(meta))

    <<
      @version::8,
      type::8,
      0::16,
      req_id::binary-size(@request_id_bytes),
      byte_size(meta_bytes)::32-big,
      meta_bytes::binary,
      byte_size(body)::32-big,
      body::binary
    >>
  end

  @doc """
  Decode a frame from binary.
  """
  @spec decode(binary()) ::
          {:ok, decoded_frame()} | {:error, :malformed_frame | :unknown_version | :unknown_type}
  def decode(
        <<@version::8, type_code::8, flags::16, request_id::binary-size(@request_id_bytes),
          meta_len::32-big, rest::binary>>
      ) do
    with {:ok, type} <- decode_type(type_code),
         true <- byte_size(rest) >= meta_len + 4,
         <<meta::binary-size(meta_len), body_len::32-big, body::binary>> <- rest,
         true <- byte_size(body) == body_len,
         {:ok, decoded_meta} <- safe_decode_json(meta) do
      {:ok,
       %{
         version: @version,
         type: type,
         flags: flags,
         request_id: request_id,
         meta: decoded_meta,
         body: body
       }}
    else
      false -> {:error, :malformed_frame}
      {:error, _} = error -> error
      _ -> {:error, :malformed_frame}
    end
  end

  def decode(<<version::8, _rest::binary>>) when version != @version do
    {:error, :unknown_version}
  end

  def decode(_), do: {:error, :malformed_frame}

  defp decode_type(@request_type), do: {:ok, :request}
  defp decode_type(@response_type), do: {:ok, :response}
  defp decode_type(@error_type), do: {:ok, :error}
  defp decode_type(@stream_open_type), do: {:ok, :stream_open}
  defp decode_type(@stream_data_type), do: {:ok, :stream_data}
  defp decode_type(@stream_resize_type), do: {:ok, :stream_resize}
  defp decode_type(@stream_close_type), do: {:ok, :stream_close}
  defp decode_type(@stream_exit_type), do: {:ok, :stream_exit}
  defp decode_type(@stream_error_type), do: {:ok, :stream_error}
  defp decode_type(@event_type), do: {:ok, :event}
  defp decode_type(_), do: {:error, :unknown_type}

  defp type_code(:request), do: @request_type
  defp type_code(:response), do: @response_type
  defp type_code(:error), do: @error_type
  defp type_code(:stream_open), do: @stream_open_type
  defp type_code(:stream_data), do: @stream_data_type
  defp type_code(:stream_resize), do: @stream_resize_type
  defp type_code(:stream_close), do: @stream_close_type
  defp type_code(:stream_exit), do: @stream_exit_type
  defp type_code(:stream_error), do: @stream_error_type
  defp type_code(:event), do: @event_type

  defp safe_decode_json(bytes) do
    case :json.decode(bytes) do
      map when is_map(map) -> {:ok, map}
      _ -> {:error, :malformed_frame}
    end
  rescue
    _ -> {:error, :malformed_frame}
  end

  defp ensure_request_id(<<_::binary-size(@request_id_bytes)>> = request_id), do: request_id

  defp ensure_request_id(other) do
    raise ArgumentError,
          "request_id must be #{@request_id_bytes} bytes, got: #{inspect(other)}"
  end
end

File: frame/frame.go
// Package frame holds the request and response frame of the ARC handler.
//
//	[1 byte]   version
//	[1 byte]   type
//	[2 bytes]  flags
//	[16 bytes] request id
//	[4 bytes]  meta length, big endian
//	[N bytes]  meta, JSON
//	[4 bytes]  body length, big endian
//	[N bytes]  body
package frame

import (
	"bytes"
	"crypto/rand"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
)

// Version is the frame version that this package writes.
const Version = 1

// RequestIDBytes is the length of a request id.
const RequestIDBytes = 16

// Type names one kind of frame.
type Type uint8

// The frame types.
const (
	Request      Type = 1
	Response     Type = 2
	Error        Type = 3
	StreamOpen   Type = 4
	StreamData   Type = 5
	StreamResize Type = 6
	StreamClose  Type = 7
	StreamExit   Type = 8
	StreamError  Type = 9
	Event        Type = 10
)

var names = map[Type]string{
	Request: "request", Response: "response", Error: "error",
	StreamOpen: "stream_open", StreamData: "stream_data", StreamResize: "stream_resize",
	StreamClose: "stream_close", StreamExit: "stream_exit", StreamError: "stream_error",
	Event: "event",
}

// String returns the name of the frame type.
func (t Type) String() string {
	if name, ok := names[t]; ok {
		return name
	}
	return fmt.Sprintf("type(%d)", uint8(t))
}

// Errors of this package.
var (
	ErrMalformed      = errors.New("frame: malformed frame")
	ErrUnknownVersion = errors.New("frame: unknown version")
	ErrUnknownType    = errors.New("frame: unknown type")
)

// Frame is one decoded frame.
type Frame struct {
	Version   int
	Type      Type
	Flags     uint16
	RequestID []byte
	Meta      map[string]any
	Body      []byte
}

// NewRequestID returns 16 random bytes.
func NewRequestID() []byte {
	id := make([]byte, RequestIDBytes)
	if _, err := rand.Read(id); err != nil {
		panic("frame: the system gave no random bytes: " + err.Error())
	}
	return id
}

// Encode writes one frame.
func Encode(frameType Type, requestID []byte, meta map[string]any, body []byte) ([]byte, error) {
	if _, ok := names[frameType]; !ok {
		return nil, ErrUnknownType
	}
	if len(requestID) != RequestIDBytes {
		return nil, fmt.Errorf("%w: the request id is %d bytes, want %d", ErrMalformed, len(requestID), RequestIDBytes)
	}
	if meta == nil {
		meta = map[string]any{}
	}

	metaBytes, err := json.Marshal(meta)
	if err != nil {
		return nil, err
	}

	out := make([]byte, 0, 4+RequestIDBytes+4+len(metaBytes)+4+len(body))
	out = append(out, Version, uint8(frameType), 0, 0)
	out = append(out, requestID...)
	out = binary.BigEndian.AppendUint32(out, uint32(len(metaBytes)))
	out = append(out, metaBytes...)
	out = binary.BigEndian.AppendUint32(out, uint32(len(body)))
	return append(out, body...), nil
}

// EncodeError writes one error frame. The code and the message go in the meta.
func EncodeError(requestID []byte, code, message string, meta map[string]any, body []byte) ([]byte, error) {
	fields := map[string]any{"code": code, "message": message}
	for key, value := range meta {
		fields[key] = value
	}
	return Encode(Error, requestID, fields, body)
}

// EncodeEvent writes one event frame. A provider sends it without a request,
// and the topic names it.
func EncodeEvent(topic string, body []byte, meta map[string]any) ([]byte, error) {
	fields := map[string]any{}
	for key, value := range meta {
		fields[key] = value
	}
	fields["topic"] = topic
	return Encode(Event, NewRequestID(), fields, body)
}

// Decode reads one frame.
func Decode(raw []byte) (*Frame, error) {
	if len(raw) < 1 {
		return nil, ErrMalformed
	}
	if raw[0] != Version {
		return nil, ErrUnknownVersion
	}
	if len(raw) < 4+RequestIDBytes+4 {
		return nil, ErrMalformed
	}

	frameType := Type(raw[1])
	if _, ok := names[frameType]; !ok {
		return nil, ErrUnknownType
	}

	flags := binary.BigEndian.Uint16(raw[2:4])
	requestID := raw[4 : 4+RequestIDBytes]
	rest := raw[4+RequestIDBytes:]

	metaLen := int(binary.BigEndian.Uint32(rest[:4]))
	rest = rest[4:]
	if len(rest) < metaLen+4 {
		return nil, ErrMalformed
	}

	metaBytes := rest[:metaLen]
	rest = rest[metaLen:]
	bodyLen := int(binary.BigEndian.Uint32(rest[:4]))
	body := rest[4:]
	if len(body) != bodyLen {
		return nil, ErrMalformed
	}

	meta, err := decodeMeta(metaBytes)
	if err != nil {
		return nil, err
	}

	return &Frame{
		Version:   Version,
		Type:      frameType,
		Flags:     flags,
		RequestID: requestID,
		Meta:      meta,
		Body:      body,
	}, nil
}

// Code returns the error code of an error frame, and the empty string for any
// other frame.
func (f *Frame) Code() string {
	code, _ := f.Meta["code"].(string)
	return code
}

// Message returns the error message of an error frame.
func (f *Frame) Message() string {
	message, _ := f.Meta["message"].(string)
	return message
}

// decodeMeta reads the meta of a frame that an unknown peer wrote. Anything
// but an object is an error.
func decodeMeta(raw []byte) (map[string]any, error) {
	decoder := json.NewDecoder(bytes.NewReader(raw))
	var meta map[string]any
	if err := decoder.Decode(&meta); err != nil || meta == nil {
		return nil, ErrMalformed
	}
	if decoder.More() {
		return nil, ErrMalformed
	}
	return meta, nil
}
